agent/queryparser/parser.go

```diff
@@ -15,26 +15,55 @@
 package queryparser
 
 import (
+	"regexp"
+	"strings"
+
+	pg_query "github.com/pganalyze/pg_query_go"
 	"github.com/pkg/errors"
 	"vitess.io/vitess/go/vt/proto/query"
 	"vitess.io/vitess/go/vt/sqlparser"
 )
 
 // MySQL parse query and return fingeprint and placeholders.
 func MySQL(q string) (string, uint32, error) {
-	normalizedQuery, _, err := sqlparser.Parse2(q)
+	statement, _, err := sqlparser.Parse2(q)
 	if err != nil {
 		return "", 0, errors.Wrap(err, "cannot parse query")
 	}
 
 	bv := make(map[string]*query.BindVariable)
-	err = sqlparser.Normalize(normalizedQuery, sqlparser.NewReservedVars("", sqlparser.GetBindvars(normalizedQuery)), bv)
+	err = sqlparser.Normalize(statement, sqlparser.NewReservedVars("", sqlparser.GetBindvars(statement)), bv)
 	if err != nil {
 		return "", 0, errors.Wrap(err, "cannot normalize query")
 	}
 
-	parsedQuery := sqlparser.NewParsedQuery(normalizedQuery)
-	bindVars := sqlparser.GetBindvars(normalizedQuery)
+	bindVars := sqlparser.GetBindvars(statement)
+
+	return sqlparser.String(statement), uint32(len(bindVars)), nil
+}
+
+// PostgreSQL parse query and return fingeprint and placeholders.
+func PostgreSQL(q string) (string, uint32, error) {
+	query, err := pg_query.Normalize(q)
+	if err != nil {
+		return "", 0, errors.Wrap(err, "cannot normalize query")
+	}
+
+	return PostgreSQLNormalized(query)
+}
+
+// PostgreSQLNormalized parse query, which is already normalized and return fingeprint and placeholders.
+func PostgreSQLNormalized(q string) (string, uint32, error) {
+	// PG 10 and above has $ as a placeholders.
+	r, err := regexp.Compile(`[\$]{1}\d`)
+	if err != nil {
+		return "", 0, errors.Wrap(err, "cannot get placeholders count")
+	}
```

Comment:
Do we actually need to compile every time?

Reply:
@rishat-ishbulatov Since its used in actions, which could be ran multiple times, its better to keep it inside. Otherwise it will be global variable and then we need mutex there. Or another solution is on your mind?

Reply:
@JiriCtvrtka https://stackoverflow.com/questions/46746988/is-regexp-matchstring-thread-safe

Reply:
@rishat-ishbulatov Good point with thread safe regexp. I did small refactor as we discussed on the Slack.

Reply:
looks good )